File: labosi/Primjeri/Igra slagalice/slagalica.cpp
/* Sveučilište u Zagrebu, Fakultet elektrotehnike i računarstva
   Umjetna inteligencija
 
   Pretraživanje prostora stanja: Igra slagalice 
   v1.0
 
   Copyright: (c) 2011 Frane Šarić <frane.saric@fer.hr>
 
   Zadatak:
   Napišite program za rješavanje slagalice proizvoljnih dimenzija 
   uporabom algoritma A* i odgovarajuće heurističke funkcije. Početna 
   pozicija slagalice neka se učitava iz datoteke (definirajte dva primjera 
   nad kojima ćete demonstirati rad svog programa). Počevši od zadane 
   pozicije, program treba pronaći i ispisati niz poteza koji dovodi do 
   rješenja, odnosno prekinuti rad ako rješenje nije pronađeno ili ako su 
   iscrpljeni računalni resursi. Dopuštenu potrošnju računalnih resursa 
   (prostornih i vremenskih, npr. broj pohranjenih čvorova i broj ukupno 
   proširenih čvorova) neka definira korisnik pri pokretanju programa.
   Detaljnije: http://www.cut-the-knot.org/pythagoras/fifteen.shtml

   Vremenska složenost opisanog rješenja je O(k * n log n), a prostorna
   O(k * n), gdje je k broj elemenata slagalice, a n broj elemenata u stablu
   pretrage.
*/
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
#include <set>
#include <iomanip>
#include <unordered_map>

using namespace std;

/* Ovo je struktura u kojoj se pamti stanje neke slagalice.
 * Pamtimo širinu i visinu slagalice te sve elemente slagalice.
 * Element sa vrijednosti 0 označava prazno polje.
 * Oko pravih elemenata slagalice dodajemo još brojeve -1 koji nam
 * olakšavaju provjeru valjanosti poteza.
 *
 * Npr. slagalica:
 * 4 _ 3
 * 2 1 5
 * pretvorila bi se u ove elemente:
 * -1 -1 -1 -1 -1
 * -1  4  0  3 -1
 * -1  2  1  5 -1
 * -1 -1 -1 -1 -1
 *
 * U samom polju e svi su elementi zapisani linearno:
 * -1, -1, -1, -1, -1, -1, 4, 0, 3, -1, -1, 2, 1, 5, -1, -1, -1, -1, -1, -1
 */
struct slagalica {
  int W, H;
  vector<int> e;
  /* Budući da se kasnije koristi asocijativno polje (struktura map) čiji je
   * ključ slagalica, a struktura map traži da svi elementi budu poredani,
   * potrebno je definirati operator <.
   *
   * U idealnom slučaju koristili bi asocijativno polje sa raspršenim
   * adresiranjem zvano unordered_map koje bi nam pružilo nešto bolje
   * performanse (O(1) umjesto O(log n)), ali malo je kompliciranije za
   * korištenje (potrebno je definirati i operator usporedbe == i vlastitu
   * "hash" funkciju).
   */
  bool operator<(const slagalica &b) const {
    return e < b.e;
  }
  /* Ispis slagalice u neki "stream". U C++-u ostream može biti npr.
   * standardni izlaz zvan cout ili npr. neka datoteka (ofstream).
   */
  friend ostream &operator<<(ostream &os, const slagalica &s) {
    for (int y = 1; y <= s.H; ++y, os << '\n') {
      for (int x = 1; x <= s.W; ++x) {
        if (int v = s.e[x + y * (s.W + 2)]) {
          cout << ' ' << setw(2) << v;
        } else {
          cout << " __";
        }
      }
    }
    return os;
  }
};

/* Malo složenija heuristička funkcija - vraća zbroj Manhattan udaljenosti
 * svakog elementa slagalice od završne pozicije tog elementa.
 */
int heuristika(const slagalica &s) {
  int rez = 0;
  for (int y = 1; y <= s.H; ++y) {
    for (int x = 1; x <= s.W; ++x) {
      if (int v = s.e[y * (s.W + 2) + x]) {
        rez += abs((v - 1) % s.W + 1 - x) + abs((v - 1) / s.W + 1 - y);
      }
    }
  }
  return rez;
}

/* Struktura u koju se spremaju svi čvorovi koji su obiđeni algoritmom A*.
 * Svi čvorovi stabla označeni su brojevima 0..n-1 (n je broj čvorova
 * stabla).
 * Svakom čvoru odgovara slagalica u tom čvoru.
 * Za svaki čvor pamti se njegov roditelj i vrijednosti funkcija F i G.
 *
 * Osim što je potrebno (za efikasnu implementaciju algoritma A*) pronaći brzo
 * koja slagalica odgovara kojem čvoru, potreban je i drugi smjer. Pomoću
 * strukture slagalica_u_id može se za zadanu slagalicu pronaći čvor u kojem
 * je ta slagalica prije viđena u vremenu O(log n).
 * Heuristička funkcija koja se koristi je monotona, pa nikada neće
 * postojati dva različita čvora u stablu sa istom slagalicom.
 */
struct stablo_pretrage {
  map<slagalica, int> slagalica_u_id;
  vector<map<slagalica, int>::iterator> id_u_slagalicu;
  unordered_map<int, int> prethodnik_cvora;
  vector<int> F, G;
};

/* Funkcija koja popuni stablo pretrage i vrati čvor u kojem je pronađeno
 * rješenje (ili -1 ako nema rješenja).
 */
int a_zvjezdica(const slagalica &pocetna, stablo_pretrage *sp) {
  /* Popis otvorenih čvorova drži se u strukturi set koja je interno
   * implementirana kao balansirano stablo. U strukturi set traženje,
   * ubacivanje i brisanje elementa operacija je koja traje O(log n)
   * vremena. Moguće je pristupiti i najmanjem elementu u vremenu O(log n).
   *
   * Ovdje se definira funkcija za usporedbu koja kaže da jedan otvoreni
   * čvor dolazi prije drugog ako je vrijednost funkcije F tog čvora manja.
   * Ukoliko su vrijednosti funkcije F jednake svejedno je potrebno
   * razlikovati elemente, pa se uzima da je manji onaj koji je prije viđen
   * (ima manji indeks).
   */
  auto usporedba = [sp](int a, int b) {
    return sp->F[a] != sp->F[b] ? sp->F[a] < sp->F[b] : a < b;
  };
  set<int, decltype(usporedba)> otvoreni(usporedba);

  /* Početna slagalica stavlja se u korijen stabla. Računaju se vrijednosti
   * funkcija F i G za početnu slagalicu.
   */
  sp->F.push_back(heuristika(pocetna));
  sp->G.push_back(0);
  /* Istovremeno se pune strukture slagalica_u_id i id_u_slagalicu. */
  sp->id_u_slagalicu.push_back(sp->slagalica_u_id.insert({pocetna, 0}).first);
  /* Tek sada je moguće ubaciti čvor 0 u listu otvorenih čvorova. */
  otvoreni.insert(0);

  while (!otvoreni.empty()) {
    /* Uzima se najmanji čvor iz liste otvorenih, tj. onaj s najmanjom
     * vrijednosti funkcije F.
     */
    int s_id = *otvoreni.begin();
    otvoreni.erase(otvoreni.begin());
    const slagalica &s = sp->id_u_slagalicu[s_id]->first;
    /* Jedino je za ciljni čvor vrijednost heuristike 0, pa ako je to slučaj
     * pronađeno je rješenje.
     */
    if (!heuristika(s)) {
      return s_id;
    }
    /* p0 označava poziciju rupe (elementa s vrijednosti 0) */
    int p0 = find(s.e.begin(), s.e.end(), 0) - s.e.begin();
    /* Pokušava se rupa zamijeniti sa jednim od četiri susjedna elementa. */
    for (int pomak : {1, -1, s.W + 2, -(s.W + 2)}) {
      /* Provjerava se da se susjedni element nalazi unutar slagalice. */
      if (s.e[p0 + pomak] >= 0) {
        slagalica nova = s;
        swap(nova.e[p0 + pomak], nova.e[p0]);

        int novi_g = sp->G[s_id] + 1, novi_f = heuristika(nova) + novi_g;
        /* Pokuša se nova slagalica ubaciti u stablo (složenost O(log n)).
         * Ako se ista slagalica već nalazi u stablu onda će varijabla i
         * pokazivati na čvor sa već postojećom slagalicom. Inače će
         * pokazivati na novoubačeni čvor.
         */
        auto i = sp->slagalica_u_id.insert({nova, sp->id_u_slagalicu.size()});
        /* Id čvora koji sadrži slagalicu "nova". */
        int nova_id = i.first->second;
        if (i.second) {
          /* Samo ako slagalica "nova" nije postojala u stablu dopunjava se
           * struktura stabla.
           */
          sp->F.push_back(0);
          sp->G.push_back(0);
          sp->id_u_slagalicu.push_back(i.first);
        } else if (novi_f < sp->F[nova_id]) {
          /* Pronađen je bolji put do postojećeg čvora u stablu, pa je
           * potrebno promijeniti vrijednost F u tom čvoru.
           * Promjena vrijednosti elementa vrši se u tri koraka (budući da
           * nije direktno podržana sturkutrom set):
           * 1. Briše se element iz liste otvorenih čvorova (O(log n)).
           * 2. Mijenja se vrijednost elementa (O(1)).
           * 3. Element se vraća u listu otvorenih čvorova (O(log n)).
           */
          otvoreni.erase(nova_id);
        } else {
          continue;
        }
        /* Spremaju se vrijednosti funkcija F i G za novi čvor koji se zatim
         * ubacuje u listu otvorenih.
         */
        sp->F[nova_id] = novi_f, sp->G[nova_id] = novi_g;
        otvoreni.insert(nova_id);
        sp->prethodnik_cvora[nova_id] = s_id;
      }
    }
  }
  return -1;
}

/* U strukturi stablo_pretrage pamte se samo roditelji, tj. prethodnici
 * čvorova. Zato je potrebno pratiti put od rješenja prema početnom čvoru i
 * ispisati ga u obrnutom redoslijedu.
 */
void ispisi_rjesenje(int id_cvora, stablo_pretrage &sp) {
  if (sp.prethodnik_cvora.count(id_cvora)) {
    ispisi_rjesenje(sp.prethodnik_cvora[id_cvora], sp);
    cout << '\n';
  }
  cout << sp.id_u_slagalicu[id_cvora]->first;
}

int main() {
  /* Učitaj slagalicu (širinu, visinu i vrijednosti svih elemenata). */
  slagalica pocetna;
  cin >> pocetna.H >> pocetna.W;
  /* Širina i visina uvećaju se za 2 i rubovi slagalice pune se
   * vrijednostima -1.
   */
  pocetna.e.resize((pocetna.H + 2) * (pocetna.W + 2), -1);
  for (int y = 1; y <= pocetna.H; ++y) {
    for (int x = 1; x <= pocetna.W; ++x) {
      cin >> pocetna.e[x + y * (pocetna.W + 2)];
    }
  }

  stablo_pretrage sp;
  int id_zavrsnog_cvora = a_zvjezdica(pocetna, &sp);
  if (id_zavrsnog_cvora >= 0) {
    cerr << "Velicina stabla pretrage: " << sp.id_u_slagalicu.size() << '\n';
    cerr << "Duljina rjesenja: " << sp.G[id_zavrsnog_cvora] << '\n';
    ispisi_rjesenje(id_zavrsnog_cvora, sp);
  }
}

File: labosi/Primjeri/Igra slagalice/slagalica.hs
-- |
-- Sveučilište u Zagrebu, Fakultet elektrotehnike i računarstva
-- Umjetna inteligencija
--
-- Pretraživanje prostora stanja: Igra slagalice 
-- v1.0
--
-- Copyright: (c) 2011 Jan Snajder <jan.snajder@fer.hr>
--
-- Zadatak:
-- Napišite program za rješavanje slagalice proizvoljnih dimenzija 
-- uporabom algoritma A* i odgovarajuće heurističke funkcije. Početna 
-- pozicija slagalice neka se učitava iz datoteke (definirajte dva primjera 
-- nad kojima ćete demonstirati rad svog programa). Počevši od zadane 
-- pozicije, program treba pronaći i ispisati niz poteza koji dovodi do 
-- rješenja, odnosno prekinuti rad ako rješenje nije pronađeno ili ako su 
-- iscrpljeni računalni resursi. Dopuštenu potrošnju računalnih resursa 
-- (prostornih i vremenskih, npr. broj pohranjenih čvorova i broj ukupno 
-- proširenih čvorova) neka definira korisnik pri pokretanju programa.
-- Detaljnije: http://www.cut-the-knot.org/pythagoras/fifteen.shtml
--
------------------------------------------------------------------------------

import Data.List
import Data.Maybe
import Data.Array
import Data.Ord
import qualified Data.Set as S
import Control.Monad
import Text.Printf

------------------------------------------------------------------------------
-- Prikaz slagalice
------------------------------------------------------------------------------

-- Podatkovna struktura za prikaz slagalice. Radi učinkovitosti, 
-- koristi se cjelobrojno dvodimenzijsko polje, a dodatno se pohranjuje i
-- pozicija praznog mjesta. Prazno mjesto u polju prikazano je
-- vrijednošću '0'.
data Puzzle = Puzzle {
  unPuzzle :: Array (Int,Int) Int, 
  hole     :: (Int,Int) }
  deriving (Eq,Ord,Show)

-- Pomoćna funkcija (omotač) za izgradnju slagalice iz ugniježđene
-- liste (ne provjerava se ispravnost liste).
puzzle :: [[Int]] -> Puzzle
puzzle xxs = Puzzle { 
  unPuzzle = listArray ((0,0),(n-1,m-1)) (concat xxs),
  hole     = (i,j) }
  where l = concat xxs
        n = length xxs
        m = length $ head xxs
        Just i = findIndex (elem 0) xxs
        Just j = findIndex (==0) (xxs!!i)

-- Pretvorba slagalice u ugniježđenu listu.
toList :: Puzzle -> [[Int]]
toList p = chunk (m+1) xs
  where (_,(_,m)) = bounds $ unPuzzle p
        xs = elems $ unPuzzle p

-- Pomoćna funkcija za grupiranje liste u podliste zadane duljine.
chunk :: Int -> [a] -> [[a]]
chunk _ [] = []
chunk n xs = ys : chunk n zs
  where (ys,zs) = splitAt n xs

-- Funkcija za ispis slagalice. Prazno mjesto ispisuje se kao bjelina.
showPuzzle :: Puzzle -> String
showPuzzle = unlines . map (concatMap print) . toList 
  where print 0 = "   "
        print x = printf "%3d" x

------------------------------------------------------------------------------
-- Funkcija sljedbenika i ispitni predikat
------------------------------------------------------------------------------

-- Pomoćna funkcija koja zamjenjuje dva elementa polja na zadanim
-- pozicijama.
arraySwap :: (Ix i) => i -> i -> Array i e -> Array i e
arraySwap i1 i2 a = a // [(i1,a!i2),(i2,a!i1)]

-- Pomiče prazno mjesto u slagalici (prazno mjesto na poziciji (i,j)
-- zamjenjuje s pločicom na poziciji (i+x,j+y)).
moveHole :: Puzzle -> (Int,Int) -> Maybe Puzzle
moveHole p (x,y)
  | inBounds  = Just $ Puzzle { 
      unPuzzle = arraySwap (i,j) (i',j') (unPuzzle p), hole = (i',j') }
  | otherwise = Nothing
  where (i,j)     = hole p
        (i',j')   = (i+x,j+y)
        (_,(n,m)) = bounds $ unPuzzle p
        inBounds  = i' >= 0 && j' >=0 && i' <= n && j' <= m

-- Funkcija za generiranje valjanih sljedbenika zadane slagalice.
nextPuzzles :: Puzzle -> [Puzzle]
nextPuzzles p = mapMaybe (moveHole p) [(-1,0),(1,0),(0,-1),(0,1)]

-- Ispitni predikat; vraća True za ciljno stanje (prazno mjesto je
-- u doljnjem desnom kutu).
isSolved :: Puzzle -> Bool
isSolved p = l == [1..length l - 1] ++ [0]
  where l = elems $ unPuzzle p

------------------------------------------------------------------------------
-- Heurističke funkcije
------------------------------------------------------------------------------

-- Heuristika 1: broj pločica koje nisu na svome mjestu (prazno mjesto
-- se ne broji). Ova heuristika dobivena je relaksacijom problema,
-- stoga je nužno konzistentna.
heuristics1 :: Puzzle -> Int
heuristics1 = 
  length . filter (==False) . zipWith (==) [1..] . init . elems . unPuzzle

-- Heuristika 2: zbroj L1-udaljenosti svake pločice od njezinog mjesta 
-- (prazno mjesto se ne broji). Ova heuristika dobivena je
-- relaksacijom problema, stoga je nužno konzistentna.
heuristics2 :: Puzzle -> Int
heuristics2 p = sum $ map (\(ij,x) -> l1 (pos x) ij) xs
  where l1 (i1,j1) (i2,j2) = abs (i1-i2) + abs (j1-j2)
        pos x = ((x-1) `div` (m+1), (x-1) `mod` (m+1))
        (_,(n,m)) = bounds $ unPuzzle p
        xs = filter ((/=0).snd) . assocs $ unPuzzle p

------------------------------------------------------------------------------
-- Algoritam A*
------------------------------------------------------------------------------

newtype Node a = Node (a,Int,[a]) deriving (Eq,Show)

instance (Ord a) => Ord (Node a) where
  --Node (_,g1,_) <= Node (_,g2,_) = g1 <= g2
  Node (s1,_,_) <= Node (s2,_,_) = s1 <= s2

-- Usmjereno pretraživanje A* s listom posjećenih stanja. Funkcija
-- uzima početno stanje, funkciju sljedbenika, ispitni predikat,
-- heurističku funkciju i maksimalno dozvoljeni broj posjećenih
-- čvorova.  Funkcija vraća listu stanja od početnog do ciljnog
-- stanja, odnosno praznu listu, ako rješenje nije pronađeno. Ako
-- dosegne maksimalno dopušten broj čvorova, funkcija vraća
-- odgovarajuću poruku. Pretpostavke: (1) heuristika je konzistentna
-- (zato pamtimo samo posjećena stanja, a ne i zatvorene čvorove), (2)
-- cijena prijelaza je konstantna.
aStar :: (Ord a) => 
  a -> (a -> [a]) -> (a -> Bool) -> (a -> Int) -> Int -> Either String [a] 
aStar s0 succ goal h maxn = step [(s0,0,[])] S.empty 1
  where step [] v n      = Right []
        step (x@(s,g,ss):xs) v n
          | n > maxn     = Left  $ "Node limit " ++ show maxn ++ " exceeded"
          | goal s       = Right $ reverse (s:ss)
          | S.member s v = step xs v n
          | otherwise    = let es   = expand x
                               open = insertManyBy (comparing f) es xs
                           in  step open (S.insert s v) (n+1)
        expand (s,g,ss) = [(t,g+1,s:ss) | t <- succ s]
        f (s,g,_) = g + h s

-- Pomoćna funkcija: ubacuje elemente prve liste u drugu poštujući
-- zadani uređaj.
insertManyBy :: (a -> a -> Ordering) -> [a] -> [a] -> [a]
insertManyBy f xs ys = foldl (flip $ insertBy f) ys xs

-- Npr.:
-- > aStar puzzle33 nextPuzzles isSolved heuristics2 200
-- NB: Uz zadano ograničenje od 200 posjećenih čvorova, ovu slagalicu nije 
-- moguće rješiti pomoću heuristics1.

-- Ispis rješenja za zadanu slagalicu, zadanu heuristiku i zadano ograničenje 
-- broja posjećenih čvorova.
solve :: Puzzle -> (Puzzle -> Int) -> Int -> IO ()
solve p h maxn = case aStar p nextPuzzles isSolved h maxn of
  Left err -> putStrLn err
  Right ps -> forM_ (zip [1..] ps) (\(i,p) -> do
                putStrLn $ "Step " ++ show i ++ ":"
                putStrLn $ showPuzzle p)

-- Npr.
-- > solve puzzle33 heuristics2 200

------------------------------------------------------------------------------
-- Primjeri slagalica
------------------------------------------------------------------------------

puzzle22 :: Puzzle
puzzle22 = puzzle
  [[2,3],
   [0,1]]

puzzle23 :: Puzzle
puzzle23 = puzzle 
 [[2,3,0],
  [4,5,1]]

puzzle33 :: Puzzle
puzzle33 = puzzle 
  [[0,8,7],
   [6,5,4],
   [3,2,1]]

puzzle44 :: Puzzle
puzzle44 = puzzle 
  [[ 5,15, 1, 7],
   [ 9, 4, 2,11],
   [10,14, 6, 3],
   [13, 0,12, 8]]

